src/riak_ql_ddl.erl

```diff
@@ -546,24 +546,98 @@ make_empty_insert_row(Mod) ->
     Positions = Mod:get_field_positions(),
     list_to_tuple(lists:duplicate(length(Positions), [])).
 
-%% If the INSERT command does not specify a list of columns
-%% the expected behaviour is that ALL columns are specified
-%% in the VALUES clause, so we insert a list of all columns
-%% for validation purposes
--spec insert_sql_columns(module(), [field_identifier()]) -> [field_identifier()].
-insert_sql_columns(Mod, []) when is_atom(Mod) ->
-    default_insert_columns(Mod);
-insert_sql_columns(_Mod, Fields) ->
-    Fields.
-
-%% Build the default column list, if none is specified
--spec default_insert_columns(module()) -> [field_identifier()].
-default_insert_columns(Mod) when is_atom(Mod) ->
-    ColPos = Mod:get_field_positions(),
-    FormatFn = fun({Col, _Pos}) when is_list(Col) ->
-                    {identifier, Col}
-               end,
-    lists:map(FormatFn, ColPos).
+any_unknown_column(_Fields, _FInMod, true) ->
+    true;
+any_unknown_column([], [], _AnyUnknown) ->
+    false;
+any_unknown_column([], _FInMod, false) ->
+    false;
+any_unknown_column([{identifier, [F]}|FieldsT], FInMod, false) ->
+    IsKnown = lists:any(fun(El) -> El =:= F end, FInMod),
+    any_unknown_column(FieldsT, FInMod, not IsKnown).
+
+%% If the INSERT command specifies only a partial list of columns, expand the
+%% list of columns to include those known by the DDL, ensuring that the values
+%% provided and those omitted align with the columns as provided by the caller.
+-spec insert_sql_columns(module(), [field_identifier()], [data_value()]) ->
+    {[field_identifier()],[[data_value()]]}.
+insert_sql_columns(Mod, Fields, Values) when is_atom(Mod) ->
+    FInMod = [ F || {[F], _I} <- Mod:get_field_positions() ],
+    case any_unknown_column(Fields, FInMod, false) of
+        true ->
+            %% will fail downstream
```

Comment:
You may be right in letting this error be handled in `riak_kv_qry:do_insert/1` (eventually in `:lookup_field_positions/2`). Detecting and reporting it here will break existing tests (which in itself is not so much of a problem), but more importantly, will render the error handling code in riak_kv_qry effectively dead -- which in turn will surely raise the question of removing it. I would certainly prefer to stick a comment here about relying on error handling code downstream and be done with it, just as you do here.